pkg/scheduler/framework/runtime/framework.go

```diff
@@ -561,33 +561,47 @@ func (f *frameworkImpl) QueueSortFunc() framework.LessFunc {
 // *Status and its code is set to non-success if any of the plugins returns
 // anything but Success. If a non-success status is returned, then the scheduling
 // cycle is aborted.
-func (f *frameworkImpl) RunPreFilterPlugins(ctx context.Context, state *framework.CycleState, pod *v1.Pod) (status *framework.Status) {
+func (f *frameworkImpl) RunPreFilterPlugins(ctx context.Context, state *framework.CycleState, pod *v1.Pod) (_ *framework.PreFilterResult, status *framework.Status) {
 	startTime := time.Now()
 	defer func() {
 		metrics.FrameworkExtensionPointDuration.WithLabelValues(preFilter, status.Code().String(), f.profileName).Observe(metrics.SinceInSeconds(startTime))
 	}()
+	var result *framework.PreFilterResult
+	var pluginWithNodes string
+	errMessage := func(pl framework.Plugin) string {
+		if pluginWithNodes == "" {
+			return fmt.Sprintf("node(s) didn't match %s", pl.Name())
+		}
+		return fmt.Sprintf("node(s) didn't match both %s and %s", pluginWithNodes, pl.Name())
+	}
 	for _, pl := range f.preFilterPlugins {
-		status = f.runPreFilterPlugin(ctx, pl, state, pod)
-		if !status.IsSuccess() {
-			status.SetFailedPlugin(pl.Name())
-			if status.IsUnschedulable() {
-				return status
+		r, s := f.runPreFilterPlugin(ctx, pl, state, pod)
+		if !s.IsSuccess() {
+			s.SetFailedPlugin(pl.Name())
+			if s.IsUnschedulable() {
+				return nil, s
 			}
-			return framework.AsStatus(fmt.Errorf("running PreFilter plugin %q: %w", pl.Name(), status.AsError())).WithFailedPlugin(pl.Name())
+			return nil, framework.AsStatus(fmt.Errorf("running PreFilter plugin %q: %w", pl.Name(), status.AsError())).WithFailedPlugin(pl.Name())
+		}
+		result = result.Merge(r)
+		if result != nil && result.NodeNames != nil && len(result.NodeNames) == 0 {
```

Comment:
I'd suggest we move the logic checking whether it's a nil or empty result to a function:

```go
func (r *PreFilterResult) EvaluateAllNodes() bool {
	return r == nil || r.NodeNames == nil
}
```

In this way, merge() and other callers can be simplified:

```go
func (p *PreFilterResult) Merge(in *PreFilterResult) *PreFilterResult {
	if p.EvaluateAllNodes() && in.EvaluateAllNodes() {
		return nil
	}

	r := PreFilterResult{}
	if p.EvaluateAllNodes() {
		r.NodeNames = sets.NewString(in.NodeNames.UnsortedList()...)
		return &r
	}
	if in.EvaluateAllNodes() {
		r.NodeNames = sets.NewString(p.NodeNames.UnsortedList()...)
		return &r
	}

	r.NodeNames = p.NodeNames.Intersection(in.NodeNames)
	return &r
}
```